device/device.go

```diff
@@ -102,40 +126,54 @@ func (d *Device) Connect(result chan<- error) {
 			return
 		}
 
-		// First of all, ensure we have a certificate
-		if err := d.ensureCertificate(); err != nil {
+		// Define a retry policy and the operation to be executed, i.e. we want to
+		// retrieve the brokerURL with an HTTP request
+		policy := d.makeRetryPolicy()
+		operation := func() error {
+			err := d.getBrokerURL()
+
+			if err != nil && d.ConnectRetry {
+				fmt.Println("Cannot retrieve brokerURL. Retrying...")
+			}
+			return err
+		}
+
+		err := backoff.Retry(operation, policy)
+		if err != nil && d.ConnectRetry {
 			if result != nil {
-				result <- err
+				errorMsg := fmt.Sprintf("Cannot establish a connection after %d attempts.", d.connectionRetryAttempts())
+				result <- errors.New(errorMsg)
 			}
 			return
 		}
 
-		brokerURL, err := d.getBrokerURL()
-		if err != nil {
+		if err != nil && !d.ConnectRetry {
 			if result != nil {
 				result <- err
 			}
 			return
 		}
 
-		if err := d.initializeMQTTClient(brokerURL); err != nil {
+		// Ensure we have a certificate
+		if err := d.ensureCertificate(); err != nil {
 			if result != nil {
 				result <- err
 			}
 			return
 		}
 
-		// Wait for the token - we're in a coroutine anyway
-		connectToken := d.m.Connect()
-		if ok := connectToken.WaitTimeout(30 * time.Second); !ok {
+		if err := d.initializeMQTTClient(d.brokerURL); err != nil {
```

Comment:
Since `brokerURL` is now saved in the struct and `initializeMQTTClient` is a method, there's no point in passing it as argument anymore: remove the parameter and just access `d.brokerURL` in `initializeMQTTClient`